net/xia/ppal_ether/main.c

```diff
@@ -162,6 +162,296 @@ static int local_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+
+	mether->host_interface = out_interface;
+	memmove(mether->ha, cfg->xfc_lladdr, cfg->xfc_lladdr_len);
+	attach_neigh_to_interface(mether);
+
+	WARN_ON(ether_rt_iops->fxid_add_locked(&bucket, xtbl,
+					       &mether->xem_common));
+	ether_rt_iops->fib_unlock(xtbl, &bucket);
+	fib_defer_dnf(dnf, ctx_ether(ctx)->net, XIDTYPE_ETHER);
+	return 0;
+
+def_upd:
+	fib_free_dnf(dnf);
+unlock_bucket:
+	ether_rt_iops->fib_unlock(xtbl, &bucket);
+	return rc;
+}
+
+static int main_delroute(struct xip_ppal_ctx *ctx,
+			 struct fib_xid_table *xtbl, struct xia_fib_config *cfg)
+{
+	struct fib_xid_ether_main *mether;
+	struct net_device *dev;
+	struct fib_xid *fxid;
+	const char *id;
+	u32 bucket;
+	int rc;
+
+	/* Check for errors in cfg. */
+	if (!cfg->xfc_dst || !cfg->xfc_odev || !cfg->xfc_lladdr ||
+	    cfg->xfc_lladdr_len != cfg->xfc_odev->addr_len)
+		return -EINVAL;
+
+	id = cfg->xfc_dst->xid_id;
+	dev = cfg->xfc_odev;
+
+	fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, id);
+	if (!fxid) {
+		rc = -ENOENT;
+		goto unlock_bucket;
+	}
+	if (fxid->fx_table_id != XRTABLE_MAIN_INDEX) {
+		rc = -EINVAL;
+		goto unlock_bucket;
+	}
+	mether = fxid_mether(fxid);
+
+	if (memcmp(mether->ha, cfg->xfc_lladdr, cfg->xfc_lladdr_len) &&
+	    !(mether->host_interface == dev)) {
+		rc = -EINVAL;
+		goto unlock_bucket;
+	}
+
+	detach_neigh_to_interface(mether);
+	ether_rt_iops->fxid_rm_locked(&bucket, xtbl, fxid);
+	fxid_free(xtbl, fxid);
+	rc = 0;
+
+unlock_bucket:
+	ether_rt_iops->fib_unlock(xtbl, &bucket);
+	return rc;
+}
+
+static int main_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+			   struct xip_ppal_ctx *ctx, struct sk_buff *skb,
+			   struct netlink_callback *cb)
+{
+	struct nlmsghdr *nlh;
+	u32 portid = NETLINK_CB(cb->skb).portid;
+	u32 seq = cb->nlh->nlmsg_seq;
+	struct rtmsg *rtm;
+	struct fib_xid_ether_main *mether = fxid_mether(fxid);
+	struct xia_xid dst;
+	struct nlattr *ha_attr;
+	struct rtnl_xia_ether_addrs *rtha;
+
+	nlh = nlmsg_put(skb, portid, seq, RTM_NEWROUTE, sizeof(*rtm),
+			NLM_F_MULTI);
+	if (!nlh)
+		return -EMSGSIZE;
+
+	rtm = nlmsg_data(nlh);
+	rtm->rtm_family = AF_XIA;
+	rtm->rtm_dst_len = sizeof(struct xia_xid);
+	rtm->rtm_src_len = 0;
+	rtm->rtm_tos = 0;
+	rtm->rtm_table = XRTABLE_MAIN_INDEX;
+	rtm->rtm_protocol = RTPROT_UNSPEC;
+	rtm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rtm->rtm_type = RTN_UNICAST;
+	rtm->rtm_flags = 0;
+
+	dst.xid_type = xtbl->fxt_ppal_type;
+	memmove(dst.xid_id, fxid->fx_xid, XIA_XID_MAX);
+	if (unlikely(nla_put(skb, RTA_DST, sizeof(dst), &dst)))
+		goto nla_put_failure;
+
+	/* Hardware addresses. */
+	ha_attr = nla_nest_start(skb, RTA_MULTIPATH);
+	if (!ha_attr)
+		goto nla_put_failure;
+
+	rtha = nla_reserve_nohdr(skb, sizeof(*rtha));
+	if (!rtha)
+		goto nla_put_failure;
+
+	rcu_read_lock();
+
+	rtha->interface_addr_len = mether->host_interface->addr_len;
+	memmove(rtha->interface_addr, mether->ha, rtha->interface_addr_len);
+	rtha->interface_index = mether->host_interface->ifindex;
```

Comment:
The XID already have the index of the host interface and the Ethernet address of the remote host, so we don't need to complicate the dump function.

This will simplify the xipconf changes.

Reply:
So this means then we don't even have a use for rtnl_xia_ether_addrs as no attribute is being added to the dump function.

Reply:
Yes, right.